lib/rdkafka/producer.rb

```diff
@@ -54,6 +59,48 @@ def initialize(native_kafka, partitioner_name)
       end
     end
 
+    # Sets alternative set of configuration details that can be set per topic
+    # @note It is not allowed to re-set the same topic config twice because of the underlying
+    #   librdkafka caching
+    # @param topic [String] The topic name
+    # @param config [Hash] config we want to use per topic basis
+    # @param config_hash [Integer] hash of the config. We expect it here instead of computing it,
+    #   because it is already computed during the retrieval attempt in the `#produce` flow.
+    def set_topic_config(topic, config, config_hash)
+      # Ensure lock on topic reference just in case
+      @native_kafka.with_inner do |inner|
+        @topics_refs_map[topic] ||= {}
+        @topics_configs[topic] ||= {}
+
+        return if @topics_configs[topic].key?(config_hash)
+
+        # If config is empty, we create an empty reference that will be used with defaults
+        rd_topic_config = if config.empty?
+                            nil
+                          else
+                            Rdkafka::Bindings.rd_kafka_topic_conf_new.tap do |topic_config|
+                              config.each do |key, value|
+                                error_buffer = FFI::MemoryPointer.from_string(" " * 256)
+                                result = Rdkafka::Bindings.rd_kafka_topic_conf_set(
+                                  topic_config,
+                                  key.to_s,
+                                  value.to_s,
+                                  error_buffer,
+                                  256
+                                )
+
+                                unless result == :config_ok
+                                  raise Config::ConfigError.new(error_buffer.read_string)
+                                end
+                              end
+                            end
+                          end
+
+        @topics_configs[topic][config_hash] = config
+        @topics_refs_map[topic][config_hash] = Bindings.rd_kafka_topic_new(inner, topic, rd_topic_config)
```

Comment:
I verified that this does not leak `rd_topic_config` even if `rd_kafka_topic_new` fails - `rd_kafka_topic_new` destroys the config even if the function errors (e.g. https://github.com/confluentinc/librdkafka/blob/a4bf4876fadff054cf729fcb26d0d31e93cff44b/src/rdkafka_topic.c#L299-L304)

Reply:
That said: does this need to raise an exception here in the case where `rd_kafka_topic_new` fails? Otherwise it'll just write `nil` to `@topics_refs_map[topic][config_hash]`

Reply:
Yeah good point. It should not write the nil. I took it "errorless" because it used to be like this. Let me alter that.

Reply:
Ok we won't fix it here becasue `rd_kafka_errno2err` is not exposed. But I can at this stage (will create a separate issue) raise generic exception when this returns null to prevent from it leaking out.

Reply:
ref https://github.com/karafka/rdkafka-ruby/issues/451

Reply:
Raising this here. In other places a subject to refactor.